src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs

```diff
@@ -2897,6 +2820,113 @@ var projectUniqueNamesForBuildIntToUpdate
             return result;
         }
 
+        private async Task GetBuildIntegratedProjectAction(BuildIntegratedProjectActionInput projectActionInput)
+        {
+            NuGetProjectAction[] nuGetProjectActions;
+
+            if (projectActionInput.PackageIdentity != null)
+            {
+                if (projectActionInput.PrimarySources == null || projectActionInput.PrimarySources.Count == 0)
+                {
+                    throw new ArgumentNullException($"Should have value in {nameof(projectActionInput.PrimarySources)} if there is value for {nameof(projectActionInput.PackageIdentity)}");
+                }
+
+                var nugetAction = NuGetProjectAction.CreateInstallProjectAction(projectActionInput.PackageIdentity, projectActionInput.PrimarySources.First(), projectActionInput.BuildIntegratedProject);
+                nuGetProjectActions = new[] { nugetAction };
+                projectActionInput.NugetProjectActionsLookup[projectActionInput.BuildIntegratedProject.MSBuildProjectPath] = nuGetProjectActions;
+            }
+            else
+            {
+                if (!projectActionInput.NugetProjectActionsLookup.ContainsKey(projectActionInput.BuildIntegratedProject.MSBuildProjectPath))
+                {
+                    throw new ArgumentNullException($"Either should have value in {nameof(projectActionInput.NugetProjectActionsLookup)} for {projectActionInput.BuildIntegratedProject.MSBuildProjectPath} or {nameof(projectActionInput.PackageIdentity)} & {nameof(projectActionInput.PrimarySources)}");
+                }
+
+                nuGetProjectActions = projectActionInput.NugetProjectActionsLookup[projectActionInput.BuildIntegratedProject.MSBuildProjectPath];
+
+                if (nuGetProjectActions.Length == 0)
+                {
+                    // Continue to next project if there are no actions for current project.
+                    return;
+                }
+            }
+
+            // Find all sources used in the project actions
+            var sources = new HashSet<SourceRepository>(
+                nuGetProjectActions.Where(action => action.SourceRepository != null)
+                    .Select(action => action.SourceRepository),
+                    new SourceRepositoryComparer());
+
+            projectActionInput.AllSources.UnionWith(sources);
```

Comment:
Is not the "AllSources" a not thread safe shared resource?

Reply:
Addressed it.